File: app/src/main/scala/com/waz/zclient/participants/fragments/SingleParticipantFragment.scala
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.participants.fragments

import android.content.Context
import android.os.Bundle
import android.support.annotation.Nullable
import android.support.v4.view.ViewPager
import android.view.{LayoutInflater, View, ViewGroup}
import android.widget.TextView
import com.waz.ZLog.ImplicitTag._
import com.waz.threading.Threading
import com.waz.utils._
import com.waz.zclient.common.controllers.{BrowserController, ThemeController, UserAccountsController}
import com.waz.zclient.conversation.ConversationController
import com.waz.zclient.pages.main.conversation.controller.IConversationScreenController
import com.waz.zclient.pages.main.pickuser.controller.IPickUserController
import com.waz.zclient.participants.{ParticipantOtrDeviceAdapter, ParticipantsController, TabbedParticipantPagerAdapter}
import com.waz.zclient.ui.views.tab.TabIndicatorLayout
import com.waz.zclient.utils.ContextUtils._
import com.waz.zclient.utils.{RichView, StringUtils}
import com.waz.zclient.views.menus.FooterMenuCallback
import com.waz.zclient.{FragmentHelper, R}

import scala.concurrent.Future

class SingleParticipantFragment extends FragmentHelper {
  import Threading.Implicits.Ui
  implicit def ctx: Context = getActivity

  import SingleParticipantFragment._

  private lazy val participantOtrDeviceAdapter = new ParticipantOtrDeviceAdapter

  private lazy val viewPager = view[ViewPager](R.id.vp_single_participant_viewpager)

  private lazy val convController         = inject[ConversationController]
  private lazy val participantsController = inject[ParticipantsController]
  private lazy val themeController        = inject[ThemeController]
  private lazy val screenController       = inject[IConversationScreenController]
  private lazy val userAccountsController = inject[UserAccountsController]
  private lazy val browserController      = inject[BrowserController]
  private lazy val pickUserController     = inject[IPickUserController]

  private lazy val userNameView = returning(view[TextView](R.id.user_name)) { vh =>
    participantsController.otherParticipantId.map(_.isDefined).onUi { visible =>
      vh.foreach(_.setVisible(visible))
    }

    participantsController.otherParticipant.map(_.getDisplayName).onUi { name =>
      vh.foreach(_.setText(name))
    }

    participantsController.otherParticipant.map(_.isVerified).onUi { visible =>
      vh.foreach { view =>
        val shield = if (visible) Option(getDrawable(R.drawable.shield_full)) else None

        shield.foreach { sh =>
          val pushDown = getDimenPx(R.dimen.wire__padding__1)
          sh.setBounds(0, pushDown, sh.getIntrinsicWidth, sh.getIntrinsicHeight + pushDown)
          view.setCompoundDrawablePadding(getDimenPx(R.dimen.wire__padding__tiny))
        }
        val old = view.getCompoundDrawables
        view.setCompoundDrawablesRelative(shield.orNull, old(1), old(2), old(3))
      }
    }
  }

  private lazy val userHandle = returning(view[TextView](R.id.user_handle)) { vh =>
    val handle = participantsController.otherParticipant.map(_.handle.map(_.string))

    handle
      .map(_.isDefined)
      .onUi(vis => vh.foreach(_.setVisible(vis)))

    handle
      .map {
        case Some(h) => StringUtils.formatHandle(h)
        case _       => ""
      }.onUi(str => vh.foreach(_.setText(str)))
  }

  override def onCreateView(inflater: LayoutInflater, viewGroup: ViewGroup, savedInstanceState: Bundle): View =
    inflater.inflate(R.layout.fragment_participants_single_tabbed, viewGroup, false)

  override def onViewCreated(v: View, @Nullable savedInstanceState: Bundle): Unit = {
    super.onViewCreated(v, savedInstanceState)

    viewPager.foreach { pager =>
      pager.setAdapter(new TabbedParticipantPagerAdapter(participantOtrDeviceAdapter, new FooterMenuCallback {

        override def onLeftActionClicked(): Unit =
          participantsController.isGroup.head.flatMap {
            case false => userAccountsController.hasCreateConvPermission.head.map {
              case true => pickUserController.showPickUser(IPickUserController.Destination.PARTICIPANTS)
              case _ => //
            }
            case _ => Future.successful {
              participantsController.onHideParticipants ! {}
              participantsController.otherParticipantId.head.foreach {
                case Some(userId) => userAccountsController.getOrCreateAndOpenConvFor(userId)
                case _ =>
              }
            }
          }

        override def onRightActionClicked(): Unit =
          (for {
            isGroup <- participantsController.isGroup.head
            convId  <- convController.currentConvId.head
          } yield (isGroup, convId)).flatMap {
            case (false, convId) => Future.successful(screenController.showConversationMenu(false, convId))
            case (true,  convId) => userAccountsController.hasRemoveConversationMemberPermission(convId).head.flatMap {
              case true =>
                participantsController.otherParticipantId.head.map {
                  case Some(userId) => participantsController.showRemoveConfirmation(userId)
                  case _ => //
                }
              case _ => Future.successful({})
            }
          }
        }))

      val tabIndicatorLayout = findById[TabIndicatorLayout](v, R.id.til_single_participant_tabs)
      tabIndicatorLayout.setPrimaryColor(getColorWithTheme(if (themeController.isDarkTheme) R.color.text__secondary_dark else R.color.text__secondary_light, getContext))
      tabIndicatorLayout.setViewPager(pager)
    }

    if (Option(savedInstanceState).isEmpty) viewPager.foreach { pager =>
      pager.setCurrentItem(getStringArg(ArgPageToOpen) match {
        case Some(TagDevices) => 1
        case _                => 0
      })

    }

    participantOtrDeviceAdapter.onClientClick.onUi { client =>
      participantsController.otherParticipantId.head.foreach {
        case Some(userId) =>
          Option(getParentFragment).foreach {
            case f: ParticipantFragment => f.showOtrClient(userId, client.id)
            case _ =>
          }
        case _ =>
      }
    }

    participantOtrDeviceAdapter.onHeaderClick {
      _ => browserController.openUrl(getString(R.string.url_otr_learn_why))
    }

    userNameView
    userHandle
  }

  override def onBackPressed(): Boolean = {
    participantsController.unselectParticipant()
    super.onBackPressed()
  }
}

object SingleParticipantFragment {
  val Tag: String = classOf[SingleParticipantFragment].getName
  val TagDevices: String = s"${classOf[SingleParticipantFragment].getName}/devices"

  private val ArgPageToOpen: String = "ARG_PAGE_TO_OPEN"

  def newInstance(pageToOpen: Option[String] = None): SingleParticipantFragment =
    returning(new SingleParticipantFragment) { f =>
      pageToOpen.foreach { p =>
        f.setArguments(returning(new Bundle){
          _.putString(ArgPageToOpen, p)
        })
      }
    }
}

File: app/src/main/java/com/waz/zclient/pages/main/participants/SingleParticipantFragment.java
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.pages.main.participants;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import com.waz.api.NetworkMode;
import com.waz.api.User;
import com.waz.model.ConvId;
import com.waz.model.UserId;
import com.waz.utils.wrappers.AndroidURIUtil;
import com.waz.zclient.BaseActivity;
import com.waz.zclient.OnBackPressedListener;
import com.waz.zclient.R;
import com.waz.zclient.common.controllers.SoundController;
import com.waz.zclient.common.controllers.ThemeController;
import com.waz.zclient.common.views.UserDetailsView;
import com.waz.zclient.controllers.confirmation.ConfirmationCallback;
import com.waz.zclient.controllers.confirmation.ConfirmationRequest;
import com.waz.zclient.controllers.confirmation.IConfirmationController;
import com.waz.zclient.controllers.confirmation.TwoButtonConfirmationCallback;
import com.waz.zclient.controllers.navigation.NavigationController;
import com.waz.zclient.conversation.ConversationController;
import com.waz.zclient.core.stores.connect.IConnectStore;
import com.waz.zclient.core.stores.conversation.ConversationChangeRequester;
import com.waz.zclient.core.stores.network.NetworkAction;
import com.waz.zclient.core.stores.singleparticipants.SingleParticipantStoreObserver;
import com.waz.zclient.pages.BaseFragment;
import com.waz.zclient.pages.main.connect.UserProfileContainer;
import com.waz.zclient.pages.main.participants.dialog.DialogLaunchMode;
import com.waz.zclient.participants.fragments.TabbedParticipantBodyFragment;
import com.waz.zclient.ui.animation.fragment.FadeAnimation;
import com.waz.zclient.ui.theme.OptionsTheme;
import com.waz.zclient.utils.ContextUtils;
import com.waz.zclient.utils.ViewUtils;
import com.waz.zclient.views.images.ImageAssetImageView;
import com.waz.zclient.views.menus.FooterMenu;
import com.waz.zclient.views.menus.FooterMenuCallback;
import timber.log.Timber;

public class SingleParticipantFragment extends BaseFragment<SingleParticipantFragment.Container> implements
                                                                                                 SingleParticipantStoreObserver,
                                                                                                 OnBackPressedListener,
                                                                                                 TabbedParticipantBodyFragment.Container,
                                                                                                 ParticipantBackbarFragment.Container {
    public static final String TAG = SingleParticipantFragment.class.getName();
    public static final String ARGUMENT_SHOWING_COMMON_USER = "ARGUMENT_SHOWING_COMMON_USER";
    private static final String SAVE_STATE_OTHER_USER_PROFILE_SCREEN_WAS_TRACKED = "SAVE_STATE_OTHER_USER_PROFILE_SCREEN_WAS_TRACKED";
    private static final String ARGUMENT_USER_REQUESTER = "ARGUMENT_USER_REQUESTER";

    private boolean goToConversationWithUser;
    private FooterMenu footerMenu;
    private boolean otherUserProfileScreenWasTracked;
    private ImageAssetImageView imageAssetImageViewProfile;

    public static SingleParticipantFragment newInstance(boolean showingCommonUser,
                                                        IConnectStore.UserRequester userRequester) {
        SingleParticipantFragment newFragment = new SingleParticipantFragment();

        Bundle args = new Bundle();
        args.putBoolean(ARGUMENT_SHOWING_COMMON_USER, showingCommonUser);
        args.putSerializable(ARGUMENT_USER_REQUESTER, userRequester);
        newFragment.setArguments(args);

        return newFragment;
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  Lifecycle
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public Animation onCreateAnimation(int transit, boolean enter, int nextAnim) {
        Animation animation = super.onCreateAnimation(transit, enter, nextAnim);

        if (getControllerFactory().getConversationScreenController().getPopoverLaunchMode() != DialogLaunchMode.AVATAR &&
            getControllerFactory().getConversationScreenController().getPopoverLaunchMode() != DialogLaunchMode.COMMON_USER) {
            int centerX = ContextUtils.getOrientationIndependentDisplayWidth(getActivity()) / 2;
            int centerY = ContextUtils.getOrientationIndependentDisplayHeight(getActivity()) / 2;
            int duration;
            int delay = 0;

            // Fade out animation when starting conversation directly with this user
            if (goToConversationWithUser && !enter) {
                goToConversationWithUser = false;
                duration = getResources().getInteger(R.integer.framework_animation_duration_medium);
                animation = new FadeAnimation(duration, 1, 0);
            } else {
                if (nextAnim != 0) {
                    if (enter) {
                        duration = getResources().getInteger(R.integer.open_profile__animation_duration);
                        delay = getResources().getInteger(R.integer.open_profile__delay);
                    } else {
                        duration = getResources().getInteger(R.integer.close_profile__animation_duration);
                    }
                    animation = new ProfileAnimation(enter, duration, delay, centerX, centerY);
                }
            }
        }
        return animation;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup viewGroup, Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_participants_single, viewGroup, false);

        Bundle args = getArguments();
        IConnectStore.UserRequester requester = null;
        if (args != null) {
            requester = (IConnectStore.UserRequester) args.getSerializable(ARGUMENT_USER_REQUESTER);
        }

        footerMenu = ViewUtils.getView(view, R.id.upm__footer);
        imageAssetImageViewProfile = ViewUtils.getView(view, R.id.iaiv__single_participant);
        imageAssetImageViewProfile.setDisplayType(ImageAssetImageView.DisplayType.CIRCLE);

        if (requester == IConnectStore.UserRequester.PARTICIPANTS) {
            footerMenu.setVisibility(View.GONE);
            imageAssetImageViewProfile.setVisibility(View.GONE);
            getChildFragmentManager().beginTransaction()
                                     .add(R.id.fl__participant__tab__container,
                                          TabbedParticipantBodyFragment.newInstance(TabbedParticipantBodyFragment.USER_PAGE()),
                                          TabbedParticipantBodyFragment.TAG())
                                     .commit();
        }

        View backgroundContainer = ViewUtils.getView(view, R.id.fl__send_connect_request__background_container);
        if (getControllerFactory().getNavigationController().getPagerPosition() == NavigationController.FIRST_PAGE ||
            (getControllerFactory().getConversationScreenController().getPopoverLaunchMode() == DialogLaunchMode.AVATAR ||
             getControllerFactory().getConversationScreenController().getPopoverLaunchMode() == DialogLaunchMode.COMMON_USER)) {
            backgroundContainer.setClickable(true);
        } else {
            backgroundContainer.setBackgroundColor(Color.TRANSPARENT);
        }

        // Hide footer until user is loaded
        footerMenu.setVisibility(View.GONE);

        if (savedInstanceState == null) {
            otherUserProfileScreenWasTracked = false;
        } else {
            otherUserProfileScreenWasTracked = savedInstanceState.getBoolean(SAVE_STATE_OTHER_USER_PROFILE_SCREEN_WAS_TRACKED);
        }

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();

        getStoreFactory().singleParticipantStore().addSingleParticipantObserver(this);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putBoolean(SAVE_STATE_OTHER_USER_PROFILE_SCREEN_WAS_TRACKED, otherUserProfileScreenWasTracked);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onStop() {
        getStoreFactory().singleParticipantStore().removeSingleParticipantObserver(this);

        super.onStop();
    }

    @Override
    public void onDestroyView() {
        if (getContainer() != null &&
            !getControllerFactory().isTornDown()) {
            getControllerFactory().getSingleImageController().clearReferences();
        }

        imageAssetImageViewProfile = null;
        footerMenu = null;
        super.onDestroyView();
    }


    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  SingleParticipantStoreObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onUserUpdated(final User user) {
        if (user == null) {
            return;
        }

        Timber.i("onUserUpdated(%s)", user.getName());

        imageAssetImageViewProfile.connectImageAsset(user.getPicture());

        // Show footer if profile is not for self user
        if (!user.isMe()) {
            if (!otherUserProfileScreenWasTracked) {
                otherUserProfileScreenWasTracked = true;
            }
            final boolean showingCommonUser = getArguments().getBoolean(ARGUMENT_SHOWING_COMMON_USER);
            final boolean ignoreRightClick = getArguments().getSerializable(ARGUMENT_USER_REQUESTER) == IConnectStore.UserRequester.CALL;
            if (ignoreRightClick) {
                footerMenu.setRightActionText("");
            } else if (showingCommonUser) {
                footerMenu.setRightActionText(getResources().getString(R.string.glyph__block));
            }

            footerMenu.setCallback(new FooterMenuCallback() {
                @Override
                public void onLeftActionClicked() {
                    getControllerFactory().getConversationScreenController().hideParticipants(true, false);

                    // Go to conversation with this user
                    goToConversationWithUser = true;
                    getContainer().dismissUserProfile();
                    Timber.i("onUserUpdated %s", user.getConversation().getId());
                    inject(ConversationController.class).selectConv(new ConvId(user.getConversation().getId()), ConversationChangeRequester.START_CONVERSATION);
                }

                @Override
                public void onRightActionClicked() {
                    if (ignoreRightClick) {
                        return;
                    }
                    if (showingCommonUser) {
                        showBlockConfirmation(user);
                    } else {
                        getStoreFactory().networkStore().doIfHasInternetOrNotifyUser(new NetworkAction() {
                            @Override
                            public void execute(NetworkMode networkMode) {
                                getContainer().showRemoveConfirmation(new UserId(user.getId()));
                            }

                            @Override
                            public void onNoNetwork() {
                                ViewUtils.showAlertDialog(getActivity(),
                                                          R.string.alert_dialog__no_network__header,
                                                          R.string.remove_from_conversation__no_network__message,
                                                          R.string.alert_dialog__confirmation,
                                                          null, true);
                            }
                        });
                    }
                }
            });
            if (getArguments().getSerializable(ARGUMENT_USER_REQUESTER) != IConnectStore.UserRequester.PARTICIPANTS) {
                footerMenu.setVisibility(View.VISIBLE);
            }
        }
    }

    private void showBlockConfirmation(final User user) {
        ConfirmationCallback callback = new TwoButtonConfirmationCallback() {
            @Override
            public void positiveButtonClicked(boolean checkboxIsSelected) {
                getStoreFactory().connectStore().blockUser(user);
                // Dismiss common user profile
                getContainer().dismissUserProfile();
            }

            @Override
            public void negativeButtonClicked() {
            }

            @Override
            public void onHideAnimationEnd(boolean confirmed, boolean canceled, boolean checkboxIsSelected) {

            }
        };
        String header = getString(R.string.confirmation_menu__block_header);
        String text = getString(R.string.confirmation_menu__block_text_with_name, user.getDisplayName());
        String confirm = getString(R.string.confirmation_menu__confirm_block);
        String cancel = getString(R.string.confirmation_menu__cancel);
        OptionsTheme optionsTheme = ((BaseActivity) getActivity()).injectJava(ThemeController.class).getThemeDependentOptionsTheme();

        ConfirmationRequest request = new ConfirmationRequest.Builder()
            .withHeader(header)
            .withMessage(text)
            .withPositiveButton(confirm)
            .withNegativeButton(cancel)
            .withConfirmationCallback(callback)
            .withWireTheme(optionsTheme)
            .build();

        getControllerFactory().getConfirmationController().requestConfirmation(request, IConfirmationController.USER_PROFILE);

        SoundController ctrl = inject(SoundController.class);
        if (ctrl != null) {
            ctrl.playAlert();
        }
    }

    @Override
    public boolean onBackPressed() {
        Timber.i("PF onBackPressed");
        getFragmentManager().popBackStackImmediate(SingleParticipantFragment.TAG, FragmentManager.POP_BACK_STACK_INCLUSIVE);
        return true;
    }

    @Override
    public void showRemoveConfirmation(UserId userId) {
        getContainer().showRemoveConfirmation(userId);
    }

    public interface Container extends UserProfileContainer {
    }
}
